pkg/resource/deploy/step_generator.go

```diff
@@ -304,8 +299,9 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 				urn = s.URN()
 			}
 			if resourcePlan, ok := sg.deployment.newPlans.get(urn); ok {
-				// If the resource is in the plan, add the operation to the plan.
 				resourcePlan.Ops = append(resourcePlan.Ops, s.Op())
+			} else if !ConstrainedTo(s.Op(), OpSame) {
+				return nil, result.Errorf("Expected a new resource plan for %v", urn)
```

Comment:
I thought even Sames got recorded to the plan. It's that untargetted resources don't get added to the plan, we should probably check against that here rather than OpSame.

Reply:
```go
		if sg.deployment.plan != nil {
			if resourcePlan, ok := sg.deployment.plan.ResourcePlans[s.URN()]; ok {
				if len(resourcePlan.Ops) == 0 {
					return nil, result.Errorf("%v is not allowed by the plan: no more steps were expected for this resource", s.Op())
				}
				constraint := resourcePlan.Ops[0]
				// We remove the Op from the list before doing the constraint check.
				// This is because we look at Ops at the end to see if any expected operations didn't attempt to happen.
				// This op has been attempted, it just might fail its constraint.
				resourcePlan.Ops = resourcePlan.Ops[1:]
				if !ConstrainedTo(s.Op(), constraint) {
					return nil, result.Errorf("%v is not allowed by the plan: this resource is constrained to %v", s.Op(), constraint)
				}
			} else {
				if !ConstrainedTo(s.Op(), OpSame) {
					return nil, result.Errorf("%v is not allowed by the plan: no steps were expected for this resource", s.Op())
				}
			}
		}
```

This logic gave me the impression that if a resource is not in the generated plan and is a same, it's ok.

Reply:
It is ok, it would just be nice to be consistent and either we always elide adding sames to plans, or we always add resource to plans if we've seen them. Having resources that we've seen, but not added into the plan feels like it'll bite us in the future.